### PR TITLE
GridCore - RowsView scrollable should have focusable element inside when column fixing is enabled

### PR DESCRIPTION
**Pull request by Tucchhaa**

### CODE REVIEW

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+export const keyboardNavigationScrollableA11yExtender = (Base: ModuleType<KeyboardNavigationController>): ModuleType<KeyboardNavigationController> => class ScrollableA11yExtender extends Base {
+  private _$firstNotFixedCell: dxElementWrapper | undefined;
+
+  private rowsViewFocusOutHandlerContext!: (event: any) => void;
+
+  init(): void {
+    super.init();
+
+    // eslint-disable-next-line max-len
+    this.rowsViewFocusOutHandlerContext = this.rowsViewFocusOutHandlerContext || this.rowsViewFocusOutHandler.bind(this);
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: Minor comment.

Maybe we can replace the ```||``` with ```??``` and split it into two lines (and remove the eslint disable comment)?
```suggestion
    this.rowsViewFocusOutHandlerContext = this.rowsViewFocusOutHandlerContext 
      ?? this.rowsViewFocusOutHandler.bind(this);
```

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import eventsEngine from '@js/events/core/events_engine';
+import { ModuleType } from '@ts/grids/grid_core/m_types';
+
+import type { KeyboardNavigationController } from './m_keyboard_navigation';
+
+// eslint-disable-next-line max-len
+export const keyboardNavigationScrollableA11yExtender = (Base: ModuleType<KeyboardNavigationController>): ModuleType<KeyboardNavigationController> => class ScrollableA11yExtender extends Base {
+  private _$firstNotFixedCell: dxElementWrapper | undefined;
+
+  private rowsViewFocusOutHandlerContext!: (event: any) => void;
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: Let's replace ```event: any``` -> ```event: Event``` here in the code below

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        columnIndex = index;
+        break;
+      }
+    }
+
+    if (columnIndex === -1) {
+      return undefined;
+    }
+
+    return this._rowsView._getCellElement(0, columnIndex) as dxElementWrapper | undefined;
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: Let's make this function shorter:

```suggestion
    const columns = this._columnsController.getVisibleColumns();
    const columnIndex = columns.findIndex(({ fixed }) => !fixed);

    return columnIndex === -1
      ? undefined
      : this._rowsView._getCellElement(0, columnIndex) as dxElementWrapper | undefined;
```

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      && !isEmptyObject(this._focusedCellPosition);
+    const isOriginalHandlerRequired = !isCellPositionDefined
+      || (!eventArgs.shift && this._isLastValidCell(this._focusedCellPosition))
+      || (eventArgs.shift && this._isFirstValidCell(this._focusedCellPosition));
+
+    if (isOriginalHandlerRequired) {
+      if (this.isScrollableNeedFocusable()) {
+        this._$firstNotFixedCell?.removeAttr('tabIndex');
+      }
+    }
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: Let's union these two conditions into one:

```suggestion
    const isNeedFocusable = this.isScrollableNeedFocusable();
    
    if (isNeedFocusable && isOriginalHandlerRequired) {
      this._$firstNotFixedCell?.removeAttr('tabIndex');
    }
```

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  private makeScrollableFocusableIfNeed(): void {
+    const needFocusable = this.isScrollableNeedFocusable();
+
+    if (!needFocusable) {
+      return;
+    }
+
+    if (this._$firstNotFixedCell) {
+      this._applyTabIndexToElement(this._$firstNotFixedCell);
+    }
```
Comment (williamvinogradov) [inline packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: Let's union these two conditions into one:

```suggestion
    if (!needFocusable || !this._$firstNotFixedCell) {
      return;
    }

    this._applyTabIndexToElement(this._$firstNotFixedCell);
```

File: packages/devextreme/testing/testcafe/helpers/accessibilityUtils.ts [hunk 6]
```diff
@@ -16,3 +16,7 @@ export const a11yCheck = async (t: TestController, rules: object = {}): Promise<
     .expect(results.violations.length === 0)
     .ok(createReport(results.violations));
 };
+
+export const reloadPage = async (t: TestController): Promise<void> => {
+  await t.eval(() => location.reload());
+};
```
Comment (williamvinogradov) [inline packages/devextreme/testing/testcafe/helpers/accessibilityUtils.ts]: Let's remove this helper function and remove ```disablePageReloads``` instead :)

If disabling ```disablePageReloads``` isn't ok for all test cases in the test file, we can move new tests to a separate file without this option.

```suggestion
```

File: packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts [hunk 7]
```diff
@@ -131,6 +132,8 @@ test('Fixed columns should have same width as not fixed columns with columnAutoW
     },
     '#otherContainer',
   );
+}).after(async () => {
+  await ClientFunction(() => { $('#otherContainer').empty(); })();
```
Comment (williamvinogradov) [inline packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts]: Question.

Do we need this if we disable page reload for these tests?

File: packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts [hunk 8]
```diff
@@ -319,3 +322,214 @@ test('Hovering over a row should work correctly after scrolling when there is a
...
+  )();
+  const columnsLength = await ClientFunction(
+    () => (window as any).widget.getVisibleColumns().length,
+  )();
+
+  const pressKey = async (key: string, times = 1) => {
+    for (let i = 0; i < times; i += 1) {
+      await t.pressKey(key);
+    }
+  };
```
Comment (williamvinogradov) [inline packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts]: Let's move this helper function on top of the test file.
As far as I can see, it used in other test cases too (duplicates)

File: packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts [hunk 9]
```diff
@@ -319,3 +322,214 @@ test('Hovering over a row should work correctly after scrolling when there is a
 
   await makeRowsViewTemplatesAsync(DATA_GRID_SELECTOR, 100);
 });
+
+test('Accessibility: Scrollable should always have focusable element when navigate through grid', async (t) => {
```
Comment (williamvinogradov) [inline packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts]: It's better to split this long test scenario into a few little scenarios that check only one thing

(You can focus on some cell immediately by clicking on it)

File: packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts (hunk 7 above)
Comment (Tucchhaa) [reply packages/devextreme/testing/testcafe/tests/dataGrid/fixedColumns.ts]: nope

File: packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts (hunk 1 above)
Comment (Tucchhaa) [reply packages/devextreme/js/__internal/grids/grid_core/keyboard_navigation/scrollable_a11y.ts]: ok

**Review by williamvinogradov — APPROVED** [review]

**Pull request merged**